File: C/canvas1.c
/*
canvas1.c

First attempt at a very simple canvas drawing app

*/

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

typedef struct {
  int r;
  int g;
  int b;
} colour;

colour white = {255, 255, 255};
colour black = {0, 0, 0};
colour red = {255, 0, 0};
colour green = {0, 255, 0};
colour blue = {0, 0, 255};

typedef struct {
  int w;
  int h;
  colour *pixels;
} image;


image * image_alloc(int, int);
void image_free(image *);
colour image_get(image *, int, int);
void image_set(image *, int, int, colour);
void image_write(image *, char *);


int main(int argc, char *argv[]) {
  image *im;
  im = image_alloc(10, 8);
  image_set(im, 3, 2, black);
  image_set(im, 4, 3, red);
  image_write(im, "test1.ppm");
}


image * image_alloc(int width, int height) {
  image * im;
  int i, j;
  im = malloc(sizeof(image));
  im->w = width;
  im->h = height;
  im->pixels = malloc(width*height*sizeof(colour));
  for (i=0;i<width;i++) {
    for (j=0;j<height;j++) {
      image_set(im, i, j, white);
    }
  }
  return(im);
}

void image_free(image * im) {
  free(im->pixels);
  free(im);
}

colour image_get(image * im, int x, int y) {
  return(im->pixels[y*(im->w) + x]);
}

void image_set(image * im, int x, int y, colour c) {
  im->pixels[y*(im->w) + x] = c;
}

// Output in the plain ASCII "P3" PPM format
// https://netpbm.sourceforge.net/doc/ppm.html
void image_write(image * im, char * fileName) {
  FILE *s;
  int i,j;
  colour c;
  s = fopen(fileName, "w");
  fprintf(s, "P3 %d %d 255\n", im->w, im->h);
  for (j=0;j<(im->h);j++) {
    for (i=0;i<(im->w);i++) {
      c = image_get(im, i, j);
      fprintf(s, "%d %d %d\n", c.r, c.g, c.b);
    }
  }
  fclose(s);
}


// eof

File: Scala/canvas1.scala
//> using scala 3.3.0

/*
canvas1.scala

Initial image/canvas app with minimal features
Creates a simple PPM

scala-cli canvas1.scala

*/


case class Colour(r: Int, g: Int, b: Int)

case class Image[T](w: Int, h: Int, data: Vector[T]):
  def apply(x: Int, y: Int): T = data(x*h+y)
  def map[S](f: T => S): Image[S] = Image(w, h, data map f)
  def updated(x: Int, y: Int, value: T): Image[T] =
    Image(w, h, data.updated(x*h+y, value))

case object Image:
  def blank(w: Int, h: Int): Image[Colour] =
    Image(w, h, Vector.fill[Colour](w*h)(Colour(255, 255, 255)))
  def saveAsPPM(im: Image[Colour], fileName: String): Unit =
    val fs = new java.io.FileWriter(fileName)
    fs.write(s"P3 ${im.w} ${im.h} 255\n")
    (0 until im.h).foreach(y =>
      (0 until im.w).foreach(x =>
        val p = im(x, y)
        fs.write(s"${p.r} ${p.g} ${p.b}\n")
      )
    )
    fs.close()

object CanvasApp:

  val white = Colour(255, 255, 255)
  val black = Colour(0, 0, 0)
  val red = Colour(255, 0, 0)
  val green = Colour(0, 255, 0)
  val blue = Colour(0, 0, 255)

  @main def main() =
    println("Hello")
    val im = Image.blank(10, 8)
    val im1 = im.
      updated(3, 2, black).
      updated(4, 3, red)
    Image.saveAsPPM(im1, "test1.ppm")
    println("Goodbye")

// eof

